sw/device/lib/testing/flash_ctrl_testutils.c

```diff
@@ -0,0 +1,148 @@
+      .size = region_size};
+
+  CHECK_DIF_OK(dif_flash_ctrl_set_data_region_properties(
+      flash_state, data_region, data_region_properties));
+  CHECK_DIF_OK(dif_flash_ctrl_set_data_region_enablement(
+      flash_state, data_region, kDifToggleEnabled));
+
+  dif_flash_ctrl_device_info_t device_info = dif_flash_ctrl_get_device_info();
+  return (base_page_index * device_info.bytes_per_page);
+}
+
+uint32_t flash_ctrl_testutils_info_region_setup(
+    dif_flash_ctrl_state_t *flash_state, uint32_t page_id) {
+  dif_flash_ctrl_region_properties_t region_properties = {
+      .ecc_en = true,
+      .high_endurance_en = false,
+      .erase_en = true,
+      .prog_en = true,
+      .rd_en = true,
+      .scramble_en = false};
+
+  dif_flash_ctrl_info_region_t info_region = {
+      .bank = 0x0, .page = page_id, .partition_id = 0x0};
+
+  CHECK_DIF_OK(dif_flash_ctrl_set_info_region_properties(
+      flash_state, info_region, region_properties));
+  CHECK_DIF_OK(dif_flash_ctrl_set_info_region_enablement(
+      flash_state, info_region, kDifToggleEnabled));
+
+  dif_flash_ctrl_device_info_t device_info = dif_flash_ctrl_get_device_info();
+  return (page_id * device_info.bytes_per_page);
+}
+
+uint32_t flash_ctrl_testutils_erase_page(
+    dif_flash_ctrl_state_t *flash_state, uint32_t byte_address,
+    dif_flash_ctrl_partition_type_t partition_type) {
+  dif_flash_ctrl_transaction_t transaction = {.byte_address = byte_address,
+                                              .op = kDifFlashCtrlOpPageErase,
+                                              .partition_type = partition_type,
+                                              .partition_id = 0x0,
+                                              .word_count = 0x0};
+  CHECK_DIF_OK(dif_flash_ctrl_start(flash_state, transaction));
+  return flash_ctrl_testutils_wait_transaction_end(flash_state);
+}
+
+uint32_t flash_ctrl_testutils_write_page(
+    dif_flash_ctrl_state_t *flash_state, uint32_t byte_address,
+    const uint32_t *data, dif_flash_ctrl_partition_type_t partition_type,
+    uint32_t word_count) {
+  dif_flash_ctrl_transaction_t transaction = {.byte_address = byte_address,
+                                              .op = kDifFlashCtrlOpProgram,
+                                              .partition_type = partition_type,
+                                              .partition_id = 0x0,
+                                              .word_count = 0x0};
+  uint32_t words_written = 0;
+  uint32_t retval = 0;
+  while (words_written < word_count) {
+    uint32_t words_to_write = ((word_count - words_written) < WORDS_PER_ACCESS)
+                                  ? (word_count - words_written)
+                                  : WORDS_PER_ACCESS;
+    transaction.word_count = words_to_write;
+    transaction.byte_address =
+        byte_address + (sizeof(uint32_t) * words_written);
+    CHECK_DIF_OK(dif_flash_ctrl_start(flash_state, transaction));
```

Comment:
@tjaychen Do we need to test the back-pressure behavior of the controller? This loop restricts transaction sizes to the what the FIFO can hold and sequences multiple transactions to perform a larger operation, rather than setting up a larger transaction and interacting with just the FIFO to complete.

Reply:
I initially tried it this way. However, it seems the flash only accepts write transactions up to a maximum number of beats per burst (which happens to be the same size as the FIFO.) After each burst of up to this maximum limit the transaction is ended and a new one started until all the data has been written. I have changed the name of the MAX constant and added a comment to make it a bit clearer.